porcupine/plugins/anchors.py

```diff
@@ -0,0 +1,150 @@
+                int(self.tab_textwidget.index(anchorpoint).split(".")[0])
+                for anchorpoint in anchor_list
+            ]
+        )
+
+        next_anchor_row = 0
+        for anchor_row in anchor_rows:
+            if anchor_row < int(cursor_row):
+                next_anchor_row = anchor_row
+            else:
+                break
+
+        if not next_anchor_row:
+            return
+        else:
+            self.tab_textwidget.mark_set("insert", f"{str(next_anchor_row)}.0")
+
+        # If cursor is above first row
+        if int(self._get_cursor_index().split(".")[0]) < int(
+            self.tab_textwidget.index("@0,0").split(".")[0]
+        ):
+            self.tab_textwidget.see("insert")
+
+        # TODO: Current Bugs:
+        # 1) Jump to previous will delete code on current row if no more anchors above.
+
+    # def bind_specific(self, event: tkinter.Event[tkinter.Misc], partial ?) -> None:
+    #     pass
+
+    # def jump_to_specific(self, event: tkinter.Event[tkinter.Misc], partial ?) -> None:
+    #     pass
+
+    def update_linenumbers(self, event: tkinter.Event[tkinter.Misc]) -> None:
+        """
+        Re-draws the anchor points every time the linenumber instance updates
+        (scroll, insertion/deletion of text)
+        """
+        self.prevent_duplicate_anchors()
+
+        anchor_list = self._get_anchors()
+
+        for anchorpoint in anchor_list:
+            row_tag = "line_" + self.tab_textwidget.index(anchorpoint).split(".")[0]
+            try:
+                [row_id] = self.linenumbers.find_withtag(row_tag)
+            except ValueError:  # if line with anchor isn't visible.
+                pass
+            else:
+                row_text = self.linenumbers.itemcget(row_id, "text")  # type: ignore[no-untyped-call]
+                self.linenumbers.itemconfigure(row_id, text=row_text + " " + self.anchor_symbol)
+
+    def prevent_duplicate_anchors(self) -> None:
+        current_rows = set()
+        anchor_list = self._get_anchors()
+
+        for anchorpoint in anchor_list:
+            anchor_row = self.tab_textwidget.index(anchorpoint).split(".")[0]
+            if anchor_row in current_rows:
+                self.tab_textwidget.mark_unset(anchorpoint)
+            else:
+                self.tab_textwidget.mark_set(anchorpoint, f"{anchorpoint} linestart")
+                current_rows.add(anchor_row)
+
+
+def on_new_tab(tab: tabs.Tab) -> None:
+    if isinstance(tab, tabs.FileTab):
+        [linenumbers] = [
+            child for child in tab.left_frame.winfo_children() if isinstance(child, LineNumbers)
+        ]
+        anchor = AnchorManager(tab.textwidget, linenumbers)
+        tab.textwidget.bind("<Control-g>", anchor.toggle_on_off, add=True)
+        tab.textwidget.bind("<Control-k>", anchor.jump_to_previous, add=True)
+        tab.textwidget.bind("<Control-l>", anchor.jump_to_next, add=True)
+        # TODO: When keybindings decided, add to default_keybindings.tcl
```

Comment:
Deciding key bindings seems to be more difficult than I thought. Ideally "jump to previous" would involve the arrow up somehow (and similarly for "jump to next" and arrow down). But:
- Up arrow by itself moves cursor up.
- Alt+Up scrolls while keeping the cursor in the center of the view
- Ctrl+Up jumps to the previous whitespace-separated section.
- Ctrl+Alt+Up doesn't do anything, but on my system, it's not possible to bind to it because Ctrl+Alt+(any arrow key) has a special meaning in general. For example, Ctrl+Alt+Right switches to the next workspace.

We get more possible key combinations by holding down shift, but it doesn't really help because it should move the cursor in the same way as without shift, but selecting as it goes. As a weird exception, Alt+Shift+Up is same as Alt+Up.

I see a couple options:
- Delete the Alt+Up binding, and use it for this instead. When I added it, I thought i would use it all the time, but I didn't end up using it very much.
- Use Alt+Shift+Up. This has the weird side effect that holding down Shift doesn't do what you would expect, but I don't think highlighting text while scrolling with cursor fixed in middle of screen would be very useful anyway.

Reply:
Forgot to mention, in case it isn't obvious yet: on Mac, you need to imagine it says "Command" whenever I wrote "Ctrl". In default_keybindings.tcl, we can use the `$contmand` variable.

Reply:
I just changed the commands to ctrl+e / r for jumping to prevent the deletion of code. However, these keybindings are only meant as temporary for developing the plugin. No thought have been put into these key bindings yet.

I don't know how you reposition your cursor per default (keyboard / mouse), for me it differs. But I think in a situation where I'm on line 5, and then jump to line 500 with an anchor, I might use the mouse to reposition the cursor to where I actually need it (not on linestart). Because of this I would keep my right hand on the mouse when using the anchors, and therefore a keybinding that could be used with one hand would make sense.

Is this a bad way to think about this?

Reply:
I wouldn't use the mouse. I would hold down control and press arrow keys :)

I think what matters the most is whether the key bindings are easy to learn. Even though "E=previous R=next" is convenient if you use a mouse, it's a bit of a gotcha compared to something that is associated with the "Up" and "Down" buttons in the obvious way. Once we use  `default_keybindings.tcl`, you can configure it however you want if you don't like the default.